packages/flame/lib/src/components/clip_component.dart

```diff
@@ -0,0 +1,92 @@
+import 'dart:ui';
+
+import 'package:flame/components.dart';
+
+enum _ClipType {
+  rect,
+  circle,
+  polygon,
+}
+
+/// {@template clip_component}
+/// A component that will clip its content.
+/// {@endtemplate}
+class ClipComponent extends PositionComponent {
+  /// {@macro clip_component}
+  ///
+  /// Creates a circle clip based on the [size].
+  ClipComponent.circle({
+    super.position,
+    super.size,
+    super.scale,
+    super.angle,
+    super.anchor,
+    super.children,
+    super.priority,
+  }) : _clipType = _ClipType.circle;
```

Comment:
Instead of this factory methods I would expect the `ClipComponent` to receive a `Clipper`. The `Clipper` would have the information on how to draw the path to be clipped. 

In this scenario, we would subclass `Clipper` and implement (example names)L`CircleClipper` `PolygonClipper` `RectClipper`.

Reply:
Could you please elaborate what would be the advantages of this approach compared to using the existing `Shape` classes?

Reply:
@st-pasha I haven't look at Flame Shape classes in a while. The above is what I would expect if I was coming from the Flutter world.

Do Shape classes allow the user to specify a custom shape?

Reply:
There are several shape classes available: circle, rectangle, rounded rectangle, and polygon. More can be added in the future.

Reply:
If the user can easily subclass a `ShapeComponent` and have the codebase live on their end, then sounds like a good idea to use `ShapeComponent` subclasses. Is there documentation on how to properly subclass ShapeComponent? For example, documentation that details how to create a StarComponent.

All I could find is: https://docs.flame-engine.org/1.2.0/flame/components.html?highlight=shapecomponent#shapecomponents

Reply:
Any chance you can plug the Shape class code URL in here?

Is this Shape class you're referencing from flame?

Reply:
I'm on my phone, just search for `shape.dart` and you should find it. (Better to ask these questions on discord so that the PR threads with the discussions don't become polluted)

Reply:
No worries @spydon are you referring to the shape in the [experimental](https://github.com/flame-engine/flame/tree/main/packages/flame/lib/src/experimental/geometry/shapes) folder?

I think we have a naming conflict with `Shape` in forge2d. So perhaps users will not have the best developer experience since they will require to "show" or "hide" or "as" imports when using flame and forge2d.

Anyways, I think using Flame's experimental Shape class is useful however I think it contains more data than is actually needed for the clipping; such as aabb. 

In addition, I don't think it would be simple, from the user POV to subclass shape or implement their own custom clipper. I believe if we go down the `Shape` road I would say the Flutter framework does a better job at exposing a simpler API to clip the canvas.

I hope my two cents help here.

Reply:
The problem though is that `Path` is very opaque, and cannot be manipulated easily. For example, if you wanted to change the shape dynamically - there isn't any way to do it. With `Shape` you can do it though: change the size of a rectangle, the radius of a rounded rectangle, or positions of individual points of a polygon.

It would be great if there was a reliable way to inspect the structure of a `Path` and modify it (or even convert `Path` into a `Shape`), but there doesn't seem to  be such a way.

Reply:
@st-pasha if the ClipComponent only required a path. Wouldn't the users be able to use the existing experimental shape classes to do all those manipulations and then use `asPath()`?

If instead, the developer wants to provide the `Shape` to later on inspect or manipulate the structure they could subclass `ClipComponent` and have a constructor that requires a `Shape` instead of a Path and then pass `Shape.asPath` to super?

Would the above solve the issue you were describing?

Reply:
Only if it functionally equivalent and provides a good API. If we go all the way with shapes it would be more difficult to provide a custom path. 

We could define a .fromShape in ClipComponent constructor that does exactly what we described above.

Reply:
A star is a `Polygon`, so that's easy. But for a more general case we ought to implement a `Polycurve` `Shape` which would be comprised of Bezier curves.
Note that with a Polygon (or Polycurve) you'd be able to manipulate individual nodes of the shape, allowing a significant degree of dynamic flexibility.

Reply:
A star was just a symbolic example. Consider instead a round edges star. 

How would a user implement a clip for a rounded edged star? Is there documentation for it? Is it just not supported with this current API? Are there examples the developer can read from to implement a custom shape?